### DISCUSSION TITLE
Can't seem to get cmd_duration working on debian (works on Arch)

### DISCUSSION THREAD

**Discussion by amituttam** [starship/starship#2039] [Q&A] [answered]
Installed starship using `cargo install starship` on a Debian machine and used AUR on an Arch box.

On Arch, I am able to get cmd_duration output. But somehow I don't see it on my Debian box. I tried forcing the cmd_duration in the `~/.config/starship.toml` but can't seem to get it show. Am I missing something obvious here? Trying other modules and settings such as gcloud, username, etc. works but not cmd_duration.

Here is my toml:

```
[gcloud]
disabled = true

[username]
style_user = "bold yellow"

[cmd_duration]
min_time = 2_000
format = "took [$duration]($style) "
style = "bold yellow"
disabled = false
show_notifications = true
min_time_to_notify = 45_000
```

Thanks.

**Comment by matchai** [comment, marked as answer]
@amituttam Which shell are you having troubles getting `cmd_duration` to show up in?
The implementation varies greatly per-shell.

**Reply by amituttam** [reply]
@matchai On Debian it is:

```
> bash --version
GNU bash, version 5.1.0(1)-rc2 (x86_64-pc-linux-gnu)
```
On Arch:

```
> bash --version
GNU bash, version 5.1.4(1)-release (x86_64-pc-linux-gnu)
```

**Reply by chipbuster** [reply]
@amituttam Just to eliminate really weird bugs, if you run `starship prompt -d 10000`, do you see timing output? (The output will be slightly mangled by bash prompt escape sequences). If not, then something is broken with starship, otherwise for some reason the shell integration isn't passing the correct information into `starship` to begin with.

**Reply by amituttam** [reply]
Here is the output.

```
Debian:
❯ starship prompt -d 10000

\[\]~\[\] on \[\]☁️ \[\]\[\]amituttam\[\] took \[\]10s\[\]
\[\]❯\[\]
```

**Reply by chipbuster** [reply]
Okay, so it looks like you're getting the time in this command. And then if you run `sleep 10` from the terminal you don't see the time? Then it's almost certainly something interfering with the timing measurement being passed to starship.

Unfortunately, I don't know how to debug further without doing some poking around. Take the following script and save it into something like ~/starship_test.sh and then modify your initscript to source it instead of doing `eval "$(starship init bash)"`. This is basically the actual initscript except with some extra lines that should tell us what starship is seeing as the `cmd_duration` argument.

If you run, e.g. `sleep 10` with this modified initscript, what do you see? I would expect that you'd see either one of "No duration in precmd" or "Found duration (some number)" every time before the prompt draws, but maybe something else happens.

```
# Will be run before *every* command (even ones in pipes!)
starship_preexec() {
    # Save previous command's last argument, otherwise it will be set to "starship_preexec"
    local PREV_LAST_ARG=$1

   echo "Running pre-exec"
    # Avoid restarting the timer for commands in the same pipeline
    if [ "$STARSHIP_PREEXEC_READY" = "true" ]; then
        STARSHIP_PREEXEC_READY=false
        STARSHIP_START_TIME=$(starship time)
        echo "Setting start time $STARSHIP_START_TIME"
    fi

    : "$PREV_LAST_ARG"
}

# Will be run before the prompt is drawn
starship_precmd() {
    local NUM_JOBS
    # Save the status, because commands in this pipeline will change $?
    STARSHIP_CMD_STATUS=$?

    # Evaluate the number of jobs before running the preseved prompt command, so that tools
    # like z/autojump, which background certain jobs, do not cause spurious background jobs
    # to be displayed by starship. Also avoids forking to run `wc`, slightly improving perf
    NUM_JOBS=$(n=0; while read line; do [[ $line ]] && n=$((n+1));done <<< $(jobs -p) ; echo $n)

    # Run the bash precmd function, if it's set. If not set, evaluates to no-op
    "${starship_precmd_user_func-:}"

    eval "$_PRESERVED_PROMPT_COMMAND"

    # Prepare the timer data, if needed.
    if [[ $STARSHIP_START_TIME ]]; then
        STARSHIP_END_TIME=$(starship time)
        STARSHIP_DURATION=$((STARSHIP_END_TIME - STARSHIP_START_TIME))
        echo "Found duration $STARSHIP_DURATION"
        PS1="$(starship prompt --status=$STARSHIP_CMD_STATUS --jobs="$NUM_JOBS" --cmd-duration=$STARSHIP_DURATION)"
        unset STARSHIP_START_TIME
    else
        echo "No duration in precmd"
        PS1="$(starship prompt --status=$STARSHIP_CMD_STATUS --jobs="$NUM_JOBS")"
    fi
    STARSHIP_PREEXEC_READY=true  # Signal that we can safely restart the timer
}

# If the user appears to be using https://github.com/rcaloras/bash-preexec,
# then hook our functions into their framework.
if [[ $preexec_functions ]]; then
    preexec_functions+=('starship_preexec "$_"')
    precmd_functions+=(starship_precmd)
else
    # We want to avoid destroying an existing DEBUG hook. If we detect one, create
    # a new function that runs both the existing function AND our function, then
    # re-trap DEBUG to use this new function. This prevents a trap clobber.
    dbg_trap="$(trap -p DEBUG | cut -d' ' -f3 | tr -d \')"
    if [[ -z "$dbg_trap" ]]; then
        trap 'starship_preexec "$_"' DEBUG
    elif [[ "$dbg_trap" != 'starship_preexec "$_"' && "$dbg_trap" != 'starship_preexec_all "$_"' ]]; then
        starship_preexec_all() {
            local PREV_LAST_ARG=$1 ; $dbg_trap; starship_preexec; : "$PREV_LAST_ARG";
        }
        trap 'starship_preexec_all "$_"' DEBUG
    fi

    # Finally, prepare the precmd function and set up the start time. We will avoid to
    # add multiple instances of the starship function and keep other user functions if any.
    if [[ -z "$PROMPT_COMMAND" ]]; then
        PROMPT_COMMAND="starship_precmd"
    elif [[ "$PROMPT_COMMAND" != *"starship_precmd"* ]]; then
        # Appending to PROMPT_COMMAND breaks exit status ($?) checking.
        # Prepending to PROMPT_COMMAND breaks "command duration" module.
        # So, we are preserving the existing PROMPT_COMMAND
        # which will be executed later in the starship_precmd function
        _PRESERVED_PROMPT_COMMAND="$PROMPT_COMMAND"
        PROMPT_COMMAND="starship_precmd"
    fi
fi

# Set up the start time and STARSHIP_SHELL, which controls shell-specific sequences
STARSHIP_START_TIME=$(starship time)
export STARSHIP_SHELL="bash"

# Set up the session key that will be used to store logs
STARSHIP_SESSION_KEY="$RANDOM$RANDOM$RANDOM$RANDOM$RANDOM"; # Random generates a number b/w 0 - 32767
STARSHIP_SESSION_KEY="${STARSHIP_SESSION_KEY}0000000000000000" # Pad it to 16+ chars.
export STARSHIP_SESSION_KEY=${STARSHIP_SESSION_KEY:0:16}; # Trim to 16-digits if excess.
```

**Reply by amituttam** [reply]
Thanks for the detailed reply.

OK so tried it. In my `.bashrc`, this is at the end:
```
#eval "$(starship init bash)"
source ~/starship_test.sh
```

With that opening a new terminal shows:
```
bash: STARSHIP_DURATION: command not found
Found duration

~
❯ sleep 10
No duration in precmd
```

**Reply by chipbuster** [reply]
Dangit, I typoed something in there. That's what I get for not testing it before posting...could you try again? I updated the script in the previous post and actually tested it locally this time.

**Reply by amituttam** [reply]
Thanks. So with this updated script, on a new terminal launch:
```
Found duration 6

~
❯ sleep 10
No duration in precmd

~
❯
```

**Reply by chipbuster** [reply]
Ugh, replied in the wrong box...

Okay, this is interesting. The preexec function isn't even firing in your case, meaning that timing information will never be generated. Just for reference, here's what I get:

```
~ 
20:57:29 ❯ sleep 10
Running pre-exec
Setting start time 1609642678296
Running pre-exec
Found duration 10008
~ took 10s 
20:58:08 ❯ 
```

 This could be because the function has somehow come untrapped or because it's firing correctly but not doing what it's supposed to.

What do you get if you just run the `trap` command in bash?

**Reply by amituttam** [reply]
Interesting. So running `trap`:
```
~
❯ trap
trap -- '__bp_preexec_invoke_exec "$_"' DEBUG
No duration in precmd
```

**Reply by chipbuster** [reply]
I'm not sure how this is happening, but your DEBUG trap is being replaced by something (or it's never being registered in the first place).

Could you try running `source ~/starship_test.sh -T` (what you're currently doing but with a `-T` flag on the source)? I've heard rumors that this can affect bash with certain builds, but I haven't seen the behavior myself.

If that doesn't fix it, what are the values of the trap (I think can be obtained just by running `trap`) right before and right after the `source ~/starship_test.sh` line in your .bashrc?

**Reply by chipbuster** [reply]
Hold up, `__bp_preexec_invoke_exec` is a bash-preexec string. You use [bash-preexec](https://github.com/rcaloras/bash-preexec)?

**Reply by amituttam** [reply]
Hmm yeah it does look like this is installed globally by our admin. I just tried on a fresh debian install on linode and it seems to work fine. What is your recommendation? It looks like the docs recommend: `Simply define the arrays preexec_functions and precmd_functions before running eval $(starship init $0), and then proceed as normal.`?

**Reply by chipbuster** [reply]
Yeah, if you're using bash-preexec and things aren't working, that's a bug on our end. Starship is supposed to seamlessly hook into bash-preexec, but it hasn't been tested in a few releases (we don't have a good way to test shell integrations in an automated manner, so sometimes rarely-used features don't get the checks they should).

I'll see if I can replicate this locally and we'll go from there.

**Reply by chipbuster** [reply]
Fails like a charm.

I'll do some debugging and open a PR to fix.

**Reply by amituttam** [reply]
OK let me know if you want me to try a patch (when ready) or run some additional tests. Regardless, happy to help further.

**Reply by chipbuster** [reply]
Could you try [this binary](https://1drv.ms/u/s!AsEbAQ2Mg0MCgfta1a5OF6Ye0j29vg?e=fcmsnG) and see if it works? SHA256 is 302c1df1c52d60dfecf09be72f1fc2f8ba3caa6360c2b868b4a10ef81a9205ab

(Unfortunately, I don't know how to make direct links in OneDrive--you'll probably need to visit the link in a browser instead of wget/curl...sorry).

**Reply by amituttam** [reply]
Do you have a patch I can try and build locally? Our policy prohibits downloading binaries, we typically have to build from source.

**Reply by chipbuster** [reply]
Sorry, I should have thought of that.

```
diff --git a/src/init/starship.bash b/src/init/starship.bash
index 2d1eb84c..c1d84ffb 100644
--- a/src/init/starship.bash
+++ b/src/init/starship.bash
@@ -56,8 +56,10 @@ starship_precmd() {
 
 # If the user appears to be using https://github.com/rcaloras/bash-preexec,
 # then hook our functions into their framework.
-if [[ $preexec_functions ]]; then
-    preexec_functions+=('starship_preexec "$_"')
+if [[ "${__bp_imported:-}" == "defined" ]]; then
+    # bash-preexec needs a single function--wrap the args into a closure and pass
+    starship_preexec_all(){ starship_preexec "$_"; }
+    preexec_functions+=(starship_preexec_all)
     precmd_functions+=(starship_precmd)
 else
     # We want to avoid destroying an existing DEBUG hook. If we detect one, create
```

**Reply by amituttam** [reply]
Just built from source with your patch, it works! Also, seems to be a lot faster when doing git status, etc...

**Reply by chipbuster** [reply]
Huh. I don't know why those commands would be faster, but I'll take it. Will open a PR with these changes once I'm back on my primary dev box. Thanks for you time and help in debugging this!

**Reply by chipbuster** [reply]
#2091 has been merged, so you should be able to use unpatched sources starting with our next release.
